src/main/java/teammates/storage/search/InstructorSearchDocument.java

```diff
@@ -0,0 +1,86 @@
+package teammates.storage.search;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrDocument;
+import org.apache.solr.common.SolrInputDocument;
+
+import teammates.common.datatransfer.attributes.CourseAttributes;
+import teammates.common.datatransfer.attributes.InstructorAttributes;
+
+/**
+ * The {@link SearchDocument} object that defines how we store {@link SolrInputDocument} for instructors.
+ */
+class InstructorSearchDocument extends SearchDocument {
+
+    private final InstructorAttributes instructor;
+
+    InstructorSearchDocument(InstructorAttributes instructor) {
+        this.instructor = instructor;
+    }
+
+    @Override
+    SolrInputDocument toDocument() {
+        SolrInputDocument document = new SolrInputDocument();
+
+        CourseAttributes course = coursesDb.getCourse(instructor.courseId);
+
+        document.addField("id", instructor.getEmail() + "%" + instructor.getCourseId());
+        document.addField("name", instructor.getName());
+        document.addField("email", instructor.getEmail());
+        document.addField("courseId", instructor.getCourseId());
+        document.addField("courseName", course == null ? "" : course.getName());
+        document.addField("googleId", instructor.getGoogleId());
+        document.addField("role", instructor.getRole());
+        document.addField("displayedName", instructor.getDisplayedName());
+
+        return document;
+    }
+
+    /**
+     * Produces a list of {@link InstructorAttributes} from the {@code QueryResponse} collection.
+     *
+     * <p>This method should be used by admin only since the searching does not restrict the
+     * visibility according to the logged-in user's google ID.</p>
+     */
+    static List<InstructorAttributes> fromResponse(QueryResponse response) {
+        if (response == null) {
+            return new ArrayList<>();
+        }
+
+        List<InstructorAttributes> instructorList = constructBaseBundle(response.getResults());
+        sortInstructorResultList(instructorList);
+
+        return instructorList;
+    }
+
+    private static List<InstructorAttributes> constructBaseBundle(List<SolrDocument> results) {
+        List<InstructorAttributes> instructorList = new ArrayList<>();
+
+        for (SolrDocument document : results) {
+            String courseId = (String) document.getFirstValue("courseId");
+            String email = (String) document.getFirstValue("email");
+            InstructorAttributes instructor = instructorsDb.getInstructorById(courseId, email);
+            if (instructor == null) {
+                // search engine out of sync as SearchManager may fail to delete documents
+                // the chance is low and it is generally not a big problem
+                instructorsDb.deleteInstructor(courseId, email);
```

Comment:
- This is deleting an instructor that we already know does not exist.
- The intention is to delete the extraneous document. Does this still achieve that?

Reply:
> The intention is to delete the extraneous document. Does this still achieve that

Right, it does not. Document deletion will be skipped when we already know entity does not exist.

Updated with deletion by id directly to remove the extraneous document.